core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java

```diff
@@ -151,6 +151,55 @@ void updateAmounts() throws AnchorException {
     Sep31Transaction txn = Context.get().getTransaction();
     Sep31PostTransactionRequest request = Context.get().getRequest();
     Amount feeResponse = Context.get().getFee();
+    boolean withoutQuote = request.getQuoteId() == null;
+
+    AssetInfo reqAsset = assetService.getAsset(request.getAssetCode(), request.getAssetIssuer());
+    BigDecimal reqAmount = decimal(request.getAmount(), reqAsset);
+    BigDecimal fee = decimal(feeResponse.getAmount(), reqAsset);
+    String feeAsset = String.valueOf(feeResponse.getAsset());
+
+    BigDecimal amountIn;
+    String amountInAsset = reqAsset.getCode();
+    BigDecimal amountOut;
+    String amountOutAsset;
+
+    if (withoutQuote) {
+      boolean strictSend = sep31Config.getPaymentType() == STRICT_SEND;
+      // Without quote.
+      amountOutAsset = amountInAsset;
+      if (strictSend) {
+        // amount_in = req.amount
+        // amount_out = amount_in - amount fee
+        amountIn = reqAmount;
+        amountOut = amountIn.subtract(fee);
+      } else {
+        // amount_in = req.amount + fee
+        // amount_out = req.amount
+        amountIn = reqAmount.add(fee);
+        amountOut = reqAmount;
+      }
+    } else {
+      // With quote
+      Sep38Quote quote = sep38QuoteStore.findByQuoteId(request.getQuoteId());
+      txn.setQuoteId(request.getQuoteId());
+      amountOutAsset = quote.getBuyAsset();
+      amountIn = decimal(quote.getSellAmount(), reqAsset);
```

Comment:
I think you should also validate that the PostTransaction amount is equals to the quote's sellAmount. If not, return a 400.

Reply:
I am not sure about this when I implement this. I actually posted the comment at [here](https://github.com/stellar/stellar-protocol/pull/1199#discussion_r870403310). 

```java
    request.amount = 100
    sep31Transaction.amount_in = 102
    sep31Transaction.amount_in_asset = USDC
    sep31Transaction.amount_out = 500
    sep31Transaction.amount_out_asset = BRL
    quote.sell_amount=102
    quote.sell_amount_asset=USDC
    quote.buy_amount=500
    quote.buy_amount_asset=BRL
```